refactor(lang/rust): rust language guide

vados-cosmonic opened the pull request:
This commit heavily refactors the rust language guide, accomplishing a few goals:

- Moving from `cargo component` to native rust toolchain
- Removing sections with advanced funtionality

**We're probably going to have to heavily discuss this, but I'm putting this PR up so we have something concrete do discuss**. 

We likely need:

1. New section(s) for the "calculator" phase/step 2 of the guide
2. New structure for the "advanced" pieces 

With regards to (1), the initial language guide seems *really* well served by the basic component + run example. It's easy to understand, and a quick win for the average new reader. Adding in imports and composition into the `calculator` component really feels like a step 2/more advanced guide that we should separate out and let people navigate to.

With regards to (2), many of the advanced sections at the bottom (type mappings, etc) really also feel like they should have their own sections. For example type mappings feels more like a reference and just doesn't make sense in the "creating components" section. 

*Maybe* this isn't the right PR for this discussion, but it really might be more fitting to rename the "creating components" section to "Build a basic component" or something along those lines. 

I'm happy to "just" convert the existing documentation 1:1 to work without cargo component, but IMO we really risk people getting lost/giving up in the middle of a very long introduction/getting started-ish page.

Resolves #319 
Resolves #207

kate-goldenring commented on component-model/src/tutorial.md:
```diff
@@ -104,51 +104,68 @@ The WebAssembly host expects it to export the [`wasi:cli/run`
-Now, implement a command line application that:
+
+> [!NOTE]
+> We have published the `docs:adder` WIT package ahead of time, so it is easy to find/access via the configuration above.
```
```suggestion
> We have published the [`docs:adder` WIT package](https://github.com/orgs/bytecodealliance/packages/container/package/docs%2Fadder) to GHCR ahead of time, so it is easy to find/access via the configuration above.
```

kate-goldenring commented on component-model/src/tutorial.md:
```diff
@@ -104,51 +104,68 @@ The WebAssembly host expects it to export the [`wasi:cli/run`
+> [!NOTE]
+> We have published the `docs:adder` WIT package ahead of time, so it is easy to find/access via the configuration above.
+>
+> To make your own custom WITs available, please use the `wkg publish` command.
```
Good opportunity to reference the wkg docs in the book
```suggestion
> To make your own custom WITs available, please use the [`wkg publish`](../../composing-and-distributing/distributing.md#distributing-wit-and-components-by-package-name-with-wkg-publish) command.
```

kate-goldenring reviewed (commented):
This is looking really close. Just some nits and help filling out a TODO

mkatychev commented on component-model/examples/tutorial/adder/src/lib.rs:
```diff
@@ -1,21 +1,27 @@
-// correlate the rust module path segments with their
-// `world.wit` counterparts:
-use bindings::exports::docs::adder::add::Guest;
-//            <- items bundled with `export` keyword
-//                     <- package namespace
-//                           <- package
-//                                  <- interface name
```
do we want to move these comments down below L23?

mkatychev replied:
looks great!

vados-cosmonic commented on component-model/src/language-support/rust.md:
```diff
@@ -1,563 +1,244 @@
-- It doesn't contain a `.wit` file. `cargo component build` will automatically export the `wasi:cli/run` interface for Rust `bin` packages, and hook it up to `main`.
-- Because there's no `.wit` file, `Cargo.toml` doesn't contain a `package.metadata.component.target` section.
-- The Rust file is called `main.rs` instead of `lib.rs`, and contains a `main` function instead of an interface implementation.
+This performs a debug build, which produces a WebAssembly module to `target/wasm32-wasip2/debug/adder.wasm`:
```
Definitely a component, thanks!

vados-cosmonic commented on component-model/src/tutorial.md:
````diff
@@ -104,51 +104,68 @@ The WebAssembly host expects it to export the [`wasi:cli/run`
 ```toml
-[package.metadata.component.target.dependencies]
-"docs:adder" = { path = "../wit/adder" }
+# TODO required wkg config
````
Ah thank you for this, this has saved me a tom of time.

mkatychev commented on component-model/src/language-support/rust.md:
````diff
@@ -1,563 +1,255 @@
+After compiling in release mode, we get a much smaller binary:
 
-## Using user-defined types
+```
+du -hs target/wasm32-wasip2/release/adder.wasm
+16K     target/wasm32-wasip2/release/adder.wasm
````
should we make this `console` since we're showing output as well?
```suggestion
```console
$ du -hs target/wasm32-wasip2/release/adder.wasm
16K     target/wasm32-wasip2/release/adder.wasm
```

mkatychev commented on component-model/examples/tutorial/adder/src/lib.rs:
```diff
@@ -1,21 +1,30 @@
-#[allow(warnings)]
-mod bindings;
+mod bindings {
+    //! This module contains generated code for implementing
+    //! the `adder` world in `wit/world.wit`.
+    //!
+    //! The `path` option is actually not required,
+    //! as by default `wit_bindgen::generate` will look
+    //! for a top-level `wit` directory and use the files
+    //! (and interfaces/worlds) there-in.
```
should we make this an outer `///` comment above `mod bindings {`? It might read cleaner with the reduced indentation.

vados-cosmonic replied:
Ah, so I think we might have a difference of opinion here -- I think the `//!` is pretty clear as it's usually what you put at the top of the file for modules... I think it's pretty clear that the `//!` applies to the module

mkatychev replied:
Makes sense to me

mkatychev commented on component-model/src/language-support/rust.md:
```diff
@@ -1,563 +1,255 @@
+> [!NOTE]
+> To use the targets above, ensure that they are enabled via the Rust toolchain (e.g. `rustup`).
+>
+> For example, to add the `wasm32-wasip2` target:
```
Would it be useful to mention the rustup subcommand below?
```suggestion
> For example, to add the `wasm32-wasip2` target (`rustup toolchain list` can be used to show all available toolchains):
```

kate-goldenring commented on component-model/src/tutorial.md:
```diff
@@ -104,51 +104,76 @@ The WebAssembly host expects it to export the [`wasi:cli/run`
+default_registry = "ghcr.io"
+
+[namespace_registries]
+# Tell wkg that packages with the `wasi` namespace are in an OCI registry under ghcr.io/webassembly 
```
While I suggested adding the wasi section, is it necessary? If so, we should explain why in the words above. If not, we should remove it or say "might as well also add support for ..."

kate-goldenring commented on component-model/src/tutorial.md:
```diff
@@ -182,3 +207,5 @@ expanded enum.
 `eval-expression` to take in a string that can then be parsed to determine which operator component
 to call. Maybe this parser is a component of its own?!
+
+[!NOTE]: #
```
is this intentional?

vados-cosmonic replied:
Yup! This is the workaround to get lint to not warn

kate-goldenring commented on component-model/src/language-support/rust.md:
```diff
@@ -1,563 +1,255 @@
+> `cargo new` so as to be sure to add `wit-bindgen` to the dependencies of the right project. Alternatively, you can directly add `wit-bindgen` to the dependencies section of the `Cargo.toml`.
 >
-> A WIT package may be spread across multiple files in the same directory; `cargo component` will search them all.
+> While it is also possible to use `wit-bindgen` as a binary via the [`wit-bindgen-cli`][crates-wit-bindgen-cli]
+> crate, but here we will focus on a code-first binding build approach.
```
```suggestion
> It is also possible to use `wit-bindgen` as a binary via the [`wit-bindgen-cli`][crates-wit-bindgen-cli]
> crate, but here we will focus on a code-first binding build approach.
```

kate-goldenring commented on component-model/src/language-support/rust.md:
````diff
@@ -1,563 +1,255 @@
 }
 ```
 
-### Implementing and exporting a resource in a component
````
Is there a reason this section on using resources was removed?

vados-cosmonic replied:
This section has now been moved, so closing this out!

kate-goldenring commented on component-model/src/language-support/rust.md:
````diff
@@ -1,563 +1,255 @@
-    ```
+After compiling in release mode, we get a much smaller binary:
 
-## Using user-defined types
````
Is there a reason this section was removed?

kate-goldenring commented on component-model/src/language-support/rust.md:
```diff
@@ -1,563 +1,255 @@
 
-> **WARNING:** If your program prints to standard out or error, you may not see the printed output! Some versions of `wasmtime` have a bug where they don't flush output streams before exiting. To work around this, add a `std::thread::sleep()` with a 10 millisecond delay before exiting `main`.
-
-### Importing an interface into a command component
```
is there a reason this section was removed?

t3hmrman replied:
Hey @kate-goldenring we're discussing this in the meeting, would love to get your thoughts here!

vados-cosmonic replied:
Hey @kate-goldenring we're discussing this in the meeting, would love to get your thoughts here!

kate-goldenring reviewed (commented):
It looks like large sections of the rust guide were removed. Is there a reason for that? Can we retain that information on how to use interfaces, resources, and WIT types?

vados-cosmonic commented on component-model/src/tutorial.md:
```diff
@@ -104,51 +104,76 @@ The WebAssembly host expects it to export the [`wasi:cli/run`
+default_registry = "ghcr.io"
+
+[namespace_registries]
+# Tell wkg that packages with the `wasi` namespace are in an OCI registry under ghcr.io/webassembly 
+wasi = { registry = "wasi",  metadata = { preferredProtocol = "oci", "oci" = {registry = "ghcr.io", namespacePrefix = "webassembly/" } } }
```
```suggestion
```

vados-cosmonic commented:
After some discussion in the meeting today, there was general alignment on splitting out the more advanced stuff in the initial guide to *other* guides. Some things it might be good to do:

- Initially, combine some advanced functionality into the *same* new section
  - Maybe we can do this in stages with progressively smaller sections split out over time? 
- If we can, have a dropdown of tutorial languages (so we avoid having 9 lines per split out tutorial)

itowlson commented:
For what it's worth I'd be opposed to removing the "advanced" stuff without having replacement content in place, because that includes widely-needed things like resources.  Happy to see those broken out into separate pages or whatever but I don't think we should just remove them with nothing in their place but an issue saying "put these back."

I appreciate I've not been following the discussion though!

vados-cosmonic commented:
Hey @kate-goldenring  @itowlson @mkatychev @macovedj thanks for the feedback so far! 

I think at this point the refactor is ready for review. It's a lot to take in, but *mostly* I tried to:

- Move the deleted guides to new sections
- Remove/update mentions of `cargo component`

I'm fairly sure some sections don't *quite* work (i.e. the directions aren't perfect), BUT I'd like to get some feedback on:

- prose
- organization
- results of adding folding (and another layer of nesting!)

Unfortunately we don't have a preview environment set up, so you'll have to build the project locally to see what the output looks like w/ folding.

Here's a preview:

<img width="1262" height="902" alt="image" src="https://github.com/user-attachments/assets/fe78e74b-0f0e-45e8-9ae5-281fcd9ef604" />

vados-cosmonic commented:
Oh also @catamorphism would love feedback here if you have some cycles!

vados-cosmonic commented on component-model/src/importing-and-reusing-components.md:
```diff
@@ -0,0 +1,26 @@
+
+## Languages
+
+This guide is implemented for various languages:
```
Ah that's a good point, this would be a good thing to add to [an issue to update the JS stuff](https://github.com/bytecodealliance/component-docs/pull/310#discussion_r2430898690)

vados-cosmonic replied:
Hey @kate-goldenring was this fine?

vados-cosmonic replied:
OK, closing this out, let's reopen if there is something else that needs to be changed

vados-cosmonic commented on component-model/src/language-support/creating-runnable-components/rust.md:
```diff
@@ -0,0 +1,161 @@
+
+> [!WARNING]
+> Reactor components can be reused, and while most platforms may *not* choose to reuse a component after `wasi:cli/run`
+> has been called, there is no guarantee that they will or will not.
```
Yeah let's remove this for now -- maybe this should go better in a FAQ down the line

Fixed in https://github.com/bytecodealliance/component-docs/pull/310/commits/0f1e1254b654a19ec557e33e154b62055b67c75b

itowlson commented on component-model/src/language-support/creating-runnable-components/rust.md:
```diff
@@ -0,0 +1,161 @@
+## Exporting the `wasi:cli/run` interface
+
+Any reactor (library-like) component can *also* export the [`run` interface][wasi-cli-iface-run] inside [WASI CLI][wasi-cli],
+and signal to ecosystem projects that it can be executed.
```
I'd suggest "to" rather than "and" but otherwise that seems fine.

I think I'm still finding the paragraph clunky because it seems like a bit of an esoteric case.  I'm guessing the motivation is like some NPM things that are both commands and libraries (helpfully, I cannot think of any right now)?  Perhaps if we lead off with motivation that will make it emerge more naturally?

vados-cosmonic commented on component-model/src/language-support/creating-runnable-components/rust.md:
```diff
@@ -0,0 +1,161 @@
+
+Building a library component this way does two things:
+
+- Enables *other* components/hosts to use the `greet` interface
```
Ah thanks, good point -- fixed in https://github.com/bytecodealliance/component-docs/pull/310/commits/cf813a422ecee3d2b5d338f1a4740d0e6d618369

vados-cosmonic commented on component-model/src/language-support/creating-runnable-components/rust.md:
```diff
@@ -0,0 +1,161 @@
+/// Component off of which implementation will hang (this can be named anything)
+struct Component;
+
+export bindings::wasi::cli::run::Guest for Component {
```
Yeah in this case the only focus here is on being runnable via `wasmtime run`, thanks to the `wasi:cli/run` export -- I'll leave a note about that in particular.

Added in https://github.com/bytecodealliance/component-docs/pull/310/commits/33dd7da7a0bf0353a75853dd5d031a87068cb21b, what do you think?

vados-cosmonic commented on component-model/src/language-support/creating-runnable-components/rust.md:
```diff
@@ -0,0 +1,161 @@
+
+## Creating a command component
+
+A _command_ is a component with a specific export that allows it to be executed directly by `wasmtime`
```
Sure, good suggestion! 

Done in https://github.com/bytecodealliance/component-docs/pull/310/commits/5e78ad4cffd4d056103a6ee3b42830cfd40fba50

Does that work for you? I added a the title and a little bit of a description on the difference/distinction.

itowlson replied:
@vados-cosmonic Yes, that's brilliant, thank you!

vados-cosmonic commented on component-model/src/language-support/creating-runnable-components/rust.md:
```diff
@@ -0,0 +1,161 @@
+> [!WARNING]
+> If your program prints to standard out or error, you may not see the printed output!
+>
+> Some versions of `wasmtime` have a bug where they don't flush output streams before exiting. To work
```
yeah I'm not sure which version of `wasmtime` had this bug and unfortunately the commit from 2 years ago didn't note the version this was on either. 

What I'll do is actually remove this though, This warning is likely no longer relevant (and this kind of bug should be filed upstream and fixed anyway if people run into it).

Done in https://github.com/bytecodealliance/component-docs/pull/310/commits/cfcfb7b7ed108fd936cf3d817235aa0ba1b1074a

vados-cosmonic commented on component-model/src/language-support/importing-and-reusing-components/rust.md:
````diff
@@ -0,0 +1,95 @@
+
+```rust
+// src/lib.rs
+mod bindings;
````
Ah thanks this is a great point! This is covered directly in the very next section ("when you build this using ..."), but I've added some notes that should clarify!

What do you think of https://github.com/bytecodealliance/component-docs/pull/310/commits/f59c26ea6387a785d14e76e7874522ca777b5968 ?

vados-cosmonic commented on component-model/src/creating-runnable-components.md:
```diff
@@ -0,0 +1,21 @@
+# Creating Runnable Components
```
So I tried expanding the existing page to note that arbitrary functions *can* be run with `wasmtime run --invoke` in https://github.com/bytecodealliance/component-docs/pull/310/commits/2c6b93e50f73b95ddca41d019028e17618c4a512, please let me know what you think!

kate-goldenring replied:
i like how you broke it up using the library/reactor vs binary/run language

kate-goldenring commented on component-model/src/language-support/creating-runnable-components/other-languages.md:
```diff
@@ -0,0 +1,20 @@
+# Creating Runnable Components (Other Languages)
+
+Just because a given programming language is not listed in this guide does not
+mean it is impossible to create runnable WebAssembly components with it.
```
"does not mean it is impossible" feels discouraging to me... not impossible but must be hard if its not here. It feels more inviting to say, "we haven't documented this yet". Suggestion also applies to other `other-language.md` pages.
```suggestion
This guide is a work in progress and does not have examples for all language toolchains with components support.
```

kate-goldenring commented on component-model/src/language-support/creating-runnable-components/other-languages.md:
```diff
@@ -0,0 +1,20 @@
+## Adding a New Language to the Guide
+
+Know of a language guide we should add to this guide? Create a PR to this [repository][repo-pr] that adds 
+the new language guide (similar to others in this section).
```
```suggestion
## Adding an Example for a Language to this Section

Are you interested in documenting this section for your language toolchain? Create a PR to this [repository][repo-pr] that adds 
the guide (similar to others in this section).
```

kate-goldenring reviewed (commented):
A few more wording changes to lift up the other languages.

vados-cosmonic commented:
Hey @kate-goldenring thanks for the review -- have addressed the comments and we should be *really close* to ready to go at least for the first version of this.

One thing I'm a little curious about is whether [building-a-simple-component/other-languages.md](https://github.com/vados-cosmonic/component-docs/blob/refactor(lang/rust)%3Drust-impl/component-model/src/language-support/building-a-simple-component/other-languages.md) should actually look like [creating-runnable-components/other-languages.md](https://github.com/vados-cosmonic/component-docs/blob/refactor(lang/rust)%3Drust-impl/component-model/src/language-support/creating-runnable-components/other-languages.md). 

I wonder if we should bother telling people to add to the list and rather just encourage them to create an issue and/or PR for their language that has support.

mkatychev commented on component-model/src/language-support/creating-runnable-components/other-languages.md:
```diff
@@ -0,0 +1,19 @@
+
+Generally, WebAssembly toolchains in the language in question may contain a way to:
+
+1. Create a WebAssembly component with the `_start` export (a "command" compnent)
```
```suggestion
1. Create a WebAssembly component with the `_start` export (a "command" component)
```

vados-cosmonic commented:
Thanks for the reviews @kate-goldenring , @mkatychev -- I've added the suggestions and made some updates (replaced the last `other-languages.md` [here](https://github.com/bytecodealliance/component-docs/pull/310/commits/1f8a7c1fbc11d8f060666d92a1c7e279100ea475))

kate-goldenring reviewed (approved):
LGTM! Thank you for all this work with this refactor! It makes it much easier to add a language guide now that you can just add a section.
